cpp/src/arrow/acero/asof_join_node.cc

```diff
@@ -548,8 +548,10 @@ class InputState {
   // true when the queue is empty and, when memo may have future entries (the case of a
   // positive tolerance), when the memo is empty.
   // used when checking whether RHS is up to date with LHS.
-  bool CurrentEmpty() const {
-    return memo_.no_future_ ? Empty() : memo_.times_.empty() && Empty();
+  // NOTE: The emptiness must be decided by an atomic all to Empty() in caller, due to the
+  // potential race with Push(), see GH-41614.
+  bool CurrentEmpty(bool empty) const {
+    return memo_.no_future_ ? empty : memo_.times_.empty() && empty;
```

Comment:
Does the `memo_.times_.empty()` value need to be obtained in a single call as well?

Reply:
Honestly I can't tell for 100% sure. But from my understanding of the code so far, we are good. Unlike the race happening on the `queue_`, between on side being `Push` and the other being `Process`, the `memo_` seems only to be manipulated by the `Process` thread which is in good sync with itself.